src/common/tusb_fifo.c

```diff
@@ -340,44 +343,47 @@ static uint16_t backward_pointer(tu_fifo_t* f, uint16_t p, uint16_t offset)
   // We limit the index space of p such that a correct wrap around happens
   // Check for a wrap around or if we are in unused index space - This has to be checked first!!
   // We are exploiting the wrap around to the correct index
-  if ((p < (uint16_t)(p - offset)) || ((uint16_t)(p - offset) > f->max_pointer_idx))
-  {
-    p = (uint16_t) ((p - offset) - f->non_used_index_space);
-  }
-  else
+  uint16_t new_p = (uint16_t) (p - offset);
+  if ( (p < new_p) || (new_p >= 2*f->depth) )
   {
-    p -= offset;
+    new_p = (uint16_t) (new_p - f->non_used_index_space);
   }
-  return p;
+
+  return new_p;
 }
 
-// get relative from absolute pointer
-static uint16_t get_relative_pointer(tu_fifo_t* f, uint16_t p)
+// index to pointer, simply an modulo with minus
+static inline uint16_t idx2ptr(uint16_t idx, uint16_t depth)
 {
-  return _ff_mod(p, f->depth);
+  while ( idx >= depth ) idx -= depth;
+  return idx;
 }
 
 // Works on local copies of w and r - return only the difference and as such can be used to determine an overflow
-static inline uint16_t _tu_fifo_count(tu_fifo_t* f, uint16_t wAbs, uint16_t rAbs)
+static inline uint16_t _tu_fifo_count(tu_fifo_t* f, uint16_t wr_idx, uint16_t rd_idx)
 {
-  uint16_t cnt = wAbs-rAbs;
+  uint16_t cnt = (uint16_t) (wr_idx-rd_idx);
 
   // In case we have non-power of two depth we need a further modification
-  if (rAbs > wAbs) cnt -= f->non_used_index_space;
+  if (rd_idx > wr_idx)
+  {
+    // 2*f->depth - (rd_idx - wr_idx);
+    cnt = (uint16_t) (cnt - f->non_used_index_space);
+  }
 
```

Comment:
How about checking here if the `cnt` value is ok?
`return (f->depth >= cnt) ? cnt : 0;`

Reply:
No you are right, i just saw it... sorry it is some time i had this code in my head by heart :D - forget this comment here please :D

Reply:
yeah, memory fades as well, and take a bit of time to revise and re-understood the fifo implementation. That is why I put a lots of effort in the header the implementation note.